runner/daemon/src/mill/daemon/MillMain0.scala

```diff
@@ -301,25 +305,83 @@ object MillMain0 {
+                          prevRunnerStateOpt,
+                          Seq("version"),
+                          initCommandLogger.streams,
+                          "BSP:initialize",
+                          loggerOpt = Some(initCommandLogger)
+                        )
 
-                      (true, RunnerState(None, Nil, None))
+                        for (err <- watchRes.error)
+                          bspLogger.streams.err.println(err)
+
+                        prevRunnerStateOpt = Some(watchRes.result)
+
+                        val sessionResultFuture = bspServerHandle.startSession(
+                          watchRes.result.frames.flatMap(_.evaluator),
+                          errored = watchRes.error.nonEmpty,
+                          watched = watchRes.watched
+                        )
+
+                        val res =
+                          if (config.bspWatch)
+                            Watching.watchAndWait(
+                              watchRes.watched,
+                              Watching.WatchArgs(
+                                setIdle = setIdle,
+                                colors = mill.internal.Colors.BlackWhite,
+                                useNotify = config.watchViaFsNotify,
+                                daemonDir = daemonDir
+                              ),
+                              () => sessionResultFuture.value,
+                              "",
+                              watchLogger.info(_)
+                            )
+                          else {
+                            watchLogger.info("Watching of build sources disabled")
+                            Some {
+                              try Success(Await.result(sessionResultFuture, Duration.Inf))
+                              catch {
+                                case NonFatal(ex) =>
+                                  Failure(ex)
+                              }
+                            }
+                          }
+
+                        // Suspend any BSP request until the next call to startSession
+                        // (that is, until we've attempted to re-compile the build)
+                        bspServerHandle.resetSession()
+
+                        res match {
+                          case None =>
+                          // Some watched meta-build files changed
+                          case Some(Failure(ex)) =>
+                            streams.err.println("BSP server threw an exception, exiting")
+                            ex.printStackTrace(streams.err)
+                            errored = true
+                            keepGoing = false
+                          case Some(Success(BspServerResult.ReloadWorkspace)) =>
+                          // reload asked by client
+                          case Some(Success(BspServerResult.Shutdown)) =>
+                            // shutdown asked by client
+                            keepGoing = false
```

Comment:
Should we log something in the BSP logs in these cases, so anyone debugging something can see the server was reloaded?

Reply:
Sure